### PR TITLE
Adding regression support

### PR DESCRIPTION
**Pull request by pcnfernando**
Implementation of AdaptiveModelRules Regressor

### CODE REVIEW

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/ClassifierPrequentialModelEvaluation.java [hunk 1]
```diff
@@ -60,9 +60,12 @@ void addResult(Instance inst, double[] classVotes) {
         }
     }
 
-    double getFractionCorrectlyClassified() {
+    //for classification
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/ClassifierPrequentialModelEvaluation.java]: Use method comments.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java [hunk 2]
```diff
@@ -182,14 +183,19 @@ private Instance createMOAInstance(double[] cepEvent) {
     }
 
     private InstancesHeader createMOAInstanceHeader(int numberOfAttributes) {
-        FastVector headerAttributes = new FastVector();
+        ArrayList<Attribute> attributes = new ArrayList<Attribute>();
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java]: List\<Attribute> attributes = new ArrayList\<Attribute>();

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java [hunk 3]
```diff
@@ -182,14 +183,19 @@ private Instance createMOAInstance(double[] cepEvent) {
...
         for (int i = 0; i < numberOfAttributes - 1; i++) {
-            headerAttributes.addElement(
-                    new Attribute("att_" + i));
+            attributes.add(new Attribute("numeric" + (i + 1)));
+        }
+        // Add class value
+        ArrayList<String> classLabels = new ArrayList<String>();
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java]: Same as above. Assign to a List.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java [hunk 4]
```diff
@@ -182,14 +183,19 @@ private Instance createMOAInstance(double[] cepEvent) {
...
+        for (int i = 0; i < this.noOfClasses; i++) {
+            classLabels.add("class" + (i + 1));
         }
-        InstancesHeader streamHeader = new InstancesHeader(new Instances
-                (this.getCLICreationString(InstanceStream.class), headerAttributes, 0));
-        streamHeader.setClassIndex(streamHeader.numAttributes());
+        attributes.add(new Attribute("class", classLabels));
+        InstancesHeader streamHeader = new InstancesHeader(new Instances(getCLICreationString(InstanceStream.class),
+                attributes, 0));
+        streamHeader.setClassIndex(noOfFeatures - 1);
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java:198]: Can you explain the logic here?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package org.wso2.extension.siddhi.gpl.execution.streamingml.regression;
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java:18]: License is missing

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import com.yahoo.labs.samoa.instances.Instance;
+import com.yahoo.labs.samoa.instances.Instances;
+import com.yahoo.labs.samoa.instances.InstancesHeader;
+import moa.options.AbstractOptionHandler;
+import moa.streams.InstanceStream;
+
+import java.util.ArrayList;
+
+/**
+ *
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java]: missing comments

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public abstract class AbstractRegressor extends AbstractOptionHandler {
+    protected InstancesHeader streamHeader;
+
+    protected void generateHeader(int noOfFeatures) {
+        ArrayList<Attribute> attributes = new ArrayList<Attribute>();
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java]: List - fix everywhere.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/Regressor.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package org.wso2.extension.siddhi.gpl.execution.streamingml.regression;
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/Regressor.java:18]: License

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
+package org.wso2.extension.siddhi.gpl.execution.streamingml.regression.adaptivemodelrules;
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java:18]: License. Please fix everywhere.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+import org.wso2.siddhi.query.api.definition.Attribute;
+import org.wso2.siddhi.query.api.exception.SiddhiAppValidationException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Do regression analysis using an AMRulesRegressor model.
+ * built via @{@link AdaptiveModelRulesRegressorStreamProcessorExtension}
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: fix the typos in the comment.

**Review by nirmal070125 — CHANGES_REQUESTED** [review]

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java (hunk 4 above)
Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java:198]: It's creating the stream header(stream definition) needed for the model. Used moa Attributes instead using a FastVector

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    private int noOfFeatures;
+    private List<VariableExpressionExecutor> featureVariableExpressionExecutors = new ArrayList<>();
+    private double[] cepEvent;
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition abstractDefinition, ExpressionExecutor[] expressionExecutors,
+                                   ConfigReader configReader, SiddhiAppContext siddhiAppContext) {
+        String modelPrefix;
+        noOfFeatures = inputDefinition.getAttributeList().size();
+
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java:108]: Need to validate the min # of arguments.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        if (attributeExpressionExecutors[0] instanceof ConstantExpressionExecutor) {
+            if (attributeExpressionExecutors[0].getReturnType() == Attribute.Type.STRING) {
+                modelPrefix = (String) ((ConstantExpressionExecutor)
+                        attributeExpressionExecutors[0])
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: formatting issue.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+                        .getValue();
+                // model name = user given name + siddhi app name
+                modelName = siddhiAppContext.getName() + "." + modelPrefix;
+            } else {
+                throw new SiddhiAppValidationException(
+                        "Invalid parameter type found for the model.name argument, "
+                                + "required " + Attribute.Type.STRING
+                                + " but found " + attributeExpressionExecutors[0].getReturnType().toString());
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: format properly.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            }
+            if (attributeExpressionLength != ((model.getNoOfFeatures()) + minNoOfParameters)) {
+                throw new SiddhiAppValidationException(String.format("Invalid number of parameters for "
+                                + "streamingml:AMRulesRegressor. This Stream Processor requires  %s "
+                                + "parameters, namely, model.name and at %s feature_attributes, "
+                                + "but found %s parameters", (minNoOfParameters + (model.getNoOfFeatures())),
+                        model.getNoOfFeatures(), attributeExpressionExecutors.length));
+            }
+
+        } catch (NullPointerException e) {
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: Why are u catching a NPE?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+        parameters = {
+                @Parameter(name = "model.name",
+                        description = "The name of the model to be used for prediction.",
+                        type = {DataType.STRING}),
+                @Parameter(name = "feature.vector",
+                        description = "The feature vector for the regression analysis",
+                        type = {DataType.INT, DataType.FLOAT, DataType.FLOAT, DataType.DOUBLE}),
+        },
+        returnAttributes = {
+                @ReturnAttribute(name = "featureVector",
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: we don't need to mention this. Let's be consistent across extensions.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    protected void process(ComplexEventChunk<StreamEvent> streamEventChunk, Processor processor,
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        synchronized (this) {
+            while (streamEventChunk.hasNext()) {
+                ComplexEvent complexEvent = streamEventChunk.next();
+                // Set feature_attributes
+                for (int i = 0; i < noOfFeatures; i++) {
+                    try {
+                        cepEvent[i] = ((Number) featureVariableExpressionExecutors.get(i)
+                                .execute(complexEvent)).doubleValue();
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java:172]: format

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+                    } catch (ClassCastException e) {
+                        throw new SiddhiAppRuntimeException(String.format("Incompatible attribute feature type"
+                                + " at position %s. Not of any numeric type. Please refer the stream definition "
+                                + "of Model[%s]", (i + 1), modelName));
+                    }
+                }
+                cepEvent[noOfFeatures] = 0;
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: default value of an int is 0. So no need to assign again.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+
+    @Override
+    public void start() {
+
+    }
+
+    @Override
+    public void stop() {
+        RegressorModelHolder.getInstance().deleteRegressorModel(modelName);
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: This shouldn't delete the model as it didn't create it.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    @Override
+    public Map<String, Object> currentState() {
+        return null;
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesRegressorStreamProcessorExtension.java]: return an empty map.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java (hunk 4 above)
Comment (nirmal070125) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java:198]: I want to no about the noOfFeatures - 1 logic.

Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/classification/hoeffdingtree/util/AdaptiveHoeffdingTreeModel.java:198]: Setting the index of the taget value for the training case

**Review by nirmal070125 — APPROVED** [review]

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java (hunk 5 above)
Comment (tishan89) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java:18]: This should be changed to GPL licence. Check all other places also

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesUpdaterStreamProcessorExtension.java [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+                )
+        }
+)
+public class AdaptiveModelRulesUpdaterStreamProcessorExtension extends StreamProcessor {
+    private static final Logger logger = Logger.getLogger(AdaptiveModelRulesUpdaterStreamProcessorExtension.class);
+
+    private static final int minNoOfFeatures = 2;
+    private static final int minNoOfParameters = 1;
+    private static final int noOfHyperParameters = 5;
+
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesUpdaterStreamProcessorExtension.java:141]: Use Uppercase for constants

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/util/AdaptiveModelRulesModel.java [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+import org.wso2.extension.siddhi.gpl.execution.streamingml.regression.AbstractRegressor;
+import org.wso2.extension.siddhi.gpl.execution.streamingml.regression.Regressor;
+import org.wso2.extension.siddhi.gpl.execution.streamingml.util.MathUtil;
+import org.wso2.siddhi.query.api.exception.SiddhiAppValidationException;
+
+/**
+ * Represents AMRules Regressor model
+ */
+public class AdaptiveModelRulesModel extends AbstractRegressor implements Regressor {
+    private static final long serialVersionUID = 1L;
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/util/AdaptiveModelRulesModel.java:42]: Use uppercase for static values

Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/util/AdaptiveModelRulesModel.java:42]: Declaring its own serialVersionUID should follow the following format. Field name should be the same.
 ANY-ACCESS-MODIFIER static final long serialVersionUID = 42L;

ref, [docs/Serializable](https://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html)

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java (hunk 5 above)
Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/AbstractRegressor.java:18]: Updated license to GPL in [#670e8f4657dbbf5dfb89bcb093bf4922d59b48d5](https://github.com/pcnfernando/siddhi-gpl-execution-streamingml/commit/670e8f4657dbbf5dfb89bcb093bf4922d59b48d5)

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesUpdaterStreamProcessorExtension.java (hunk 20 above)
Comment (pcnfernando) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/regression/adaptivemodelrules/AdaptiveModelRulesUpdaterStreamProcessorExtension.java:141]: Fixed in [#8b1af77d152d902dac4e2c1b4f2cc74e9251c74f](https://github.com/wso2-extensions/siddhi-gpl-execution-streamingml/pull/16/commits/8b1af77d152d902dac4e2c1b4f2cc74e9251c74f)

**Pull request merged**